src/libServer/LookupServer.cpp

```diff
@@ -1012,31 +1012,39 @@ Json::Value LookupServer::GetSmartContracts(const string& address) {
 
   try {
     Address addr{ToBase16AddrHelper(address)};
+    uint64_t nonce = 0;
+    {
+      shared_lock<shared_timed_mutex> lock(
+          AccountStore::GetInstance().GetPrimaryMutex());
 
-    shared_lock<shared_timed_mutex> lock(
-        AccountStore::GetInstance().GetPrimaryMutex());
-
-    const Account* account = AccountStore::GetInstance().GetAccount(addr, true);
+      const Account* account =
+          AccountStore::GetInstance().GetAccount(addr, true);
 
-    if (account == nullptr) {
-      throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
-                             "Address does not exist");
-    }
-    if (account->isContract()) {
-      throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
-                             "A contract account queried");
+      if (account == nullptr) {
+        throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
+                               "Address does not exist");
+      }
+      if (account->isContract()) {
+        throw JsonRpcException(RPC_INVALID_ADDRESS_OR_KEY,
+                               "A contract account queried");
+      }
+      nonce = account->GetNonce();
     }
-    uint64_t nonce = account->GetNonce();
+
     //[TODO] find out a more efficient way (using storage)
     Json::Value _json;
 
     for (uint64_t i = 0; i < nonce; i++) {
       Address contractAddr = Account::GetAddressForContract(addr, i);
-      const Account* contractAccount =
-          AccountStore::GetInstance().GetAccount(contractAddr, true);
+      {
+        shared_lock<shared_timed_mutex> lock(
```

Comment:
With moving mutex inside the for loop, we are expecting that if it's released even once in for loop continuation, then we would get a chance to proceed with other execution flow of `DeserializeDelta`, and then once the `DeserializeDelta` is finished, the API execution can carry on. Right ?

Reply:
Right